File: surface/widgets/watchkit/components/widgets/concepts/governance-execution/TimelockCountdown.swift
import SwiftUI

// State machine: running | warning | critical | expired | executing | completed | paused
enum TimelockCountdownWatchState {
    case running, warning, critical, expired, executing, completed, paused
}

enum TimelockCountdownWatchEvent {
    case tick, warningThreshold, criticalThreshold, expire, execute, reset, executeComplete, executeError, pause, resume
}

func timelockCountdownWatchReduce(_ state: TimelockCountdownWatchState, _ event: TimelockCountdownWatchEvent) -> TimelockCountdownWatchState {
    switch state {
    case .running:
        switch event {
        case .warningThreshold: return .warning
        case .expire: return .expired
        case .pause: return .paused
        default: return state
        }
    case .warning:
        switch event {
        case .criticalThreshold: return .critical
        case .expire: return .expired
        default: return state
        }
    case .critical:
        if case .expire = event { return .expired }
        return state
    case .expired:
        switch event {
        case .execute: return .executing
        case .reset: return .running
        default: return state
        }
    case .executing:
        switch event {
        case .executeComplete: return .completed
        case .executeError: return .expired
        default: return state
        }
    case .completed:
        return state
    case .paused:
        if case .resume = event { return .running }
        return state
    }
}

struct TimelockCountdownWatchView: View {
    let phase: String
    let deadline: Date
    let elapsed: Double
    let total: Double
    var showChallenge: Bool = true
    var warningThreshold: Double = 0.8
    var criticalThreshold: Double = 0.95
    var onExecute: (() -> Void)? = nil
    var onChallenge: (() -> Void)? = nil

    @State private var state: TimelockCountdownWatchState = .running
    @State private var timeRemaining: TimeInterval = 0

    private var progress: Double {
        total > 0 ? min(1, max(0, elapsed / total)) : 0
    }

    private var displayPhase: String {
        switch state {
        case .expired: return "Ready to execute"
        case .executing: return "Executing..."
        case .completed: return "Complete"
        case .paused: return "\(phase) (paused)"
        default: return phase
        }
    }

    private var urgencyColor: Color {
        switch state {
        case .running: return .blue
        case .warning: return .yellow
        case .critical, .expired: return .red
        case .executing: return .orange
        case .completed: return .green
        case .paused: return .secondary
        }
    }

    private var countdownText: String {
        if timeRemaining <= 0 { return "0s" }
        let total = Int(timeRemaining)
        let d = total / 86400
        let h = (total % 86400) / 3600
        let m = (total % 3600) / 60
        let s = total % 60
        var parts: [String] = []
        if d > 0 { parts.append("\(d)d") }
        if h > 0 { parts.append("\(h)h") }
        if m > 0 { parts.append("\(m)m") }
        parts.append("\(s)s")
        return parts.joined(separator: " ")
    }

    var body: some View {
        VStack(spacing: 6) {
            // Phase label
            Text(displayPhase)
                .font(.caption2)
                .fontWeight(.semibold)

            // Countdown - prominent on watch
            Text(state == .completed ? "Done" : countdownText)
                .font(.system(size: 20, weight: .bold, design: .monospaced))
                .foregroundColor(urgencyColor)

            // Progress gauge
            Gauge(value: progress) {
                EmptyView()
            }
            .gaugeStyle(.linearCapacity)
            .tint(urgencyColor)

            // Execute button
            if state == .expired {
                Button("Execute") {
                    state = timelockCountdownWatchReduce(state, .execute)
                    onExecute?()
                }
                .font(.caption2)
                .buttonStyle(.borderedProminent)
            }

            // Challenge button
            if showChallenge && state != .expired && state != .completed && state != .executing {
                Button("Challenge") {
                    onChallenge?()
                }
                .font(.caption2)
                .buttonStyle(.bordered)
            }
        }
        .onAppear { updateTimeRemaining() }
        .accessibilityElement(children: .combine)
        .accessibilityLabel("\(displayPhase): \(countdownText)")
    }

    private func updateTimeRemaining() {
        timeRemaining = max(0, deadline.timeIntervalSinceNow)
        if timeRemaining <= 0 && state != .expired && state != .executing && state != .completed {
            state = timelockCountdownWatchReduce(state, .expire)
        }
    }
}

File: surface/widgets/vue/components/widgets/concepts/governance-execution/TimelockCountdown.ts
import { defineComponent, h, ref, computed, onMounted, onUnmounted, watch } from 'vue';

/* ---------------------------------------------------------------------------
 * TimelockCountdown state machine
 * ------------------------------------------------------------------------- */

export type TimelockCountdownState = 'running' | 'warning' | 'critical' | 'expired' | 'executing' | 'completed' | 'paused';
export type TimelockCountdownEvent =
  | { type: 'TICK' }
  | { type: 'WARNING_THRESHOLD' }
  | { type: 'EXPIRE' }
  | { type: 'PAUSE' }
  | { type: 'CRITICAL_THRESHOLD' }
  | { type: 'EXECUTE' }
  | { type: 'RESET' }
  | { type: 'EXECUTE_COMPLETE' }
  | { type: 'EXECUTE_ERROR' }
  | { type: 'RESUME' }
  | { type: 'CHALLENGE' };

export function timelockCountdownReducer(state: TimelockCountdownState, event: TimelockCountdownEvent): TimelockCountdownState {
  switch (state) {
    case 'running':
      if (event.type === 'TICK') return 'running';
      if (event.type === 'WARNING_THRESHOLD') return 'warning';
      if (event.type === 'EXPIRE') return 'expired';
      if (event.type === 'PAUSE') return 'paused';
      return state;
    case 'warning':
      if (event.type === 'TICK') return 'warning';
      if (event.type === 'CRITICAL_THRESHOLD') return 'critical';
      if (event.type === 'EXPIRE') return 'expired';
      return state;
    case 'critical':
      if (event.type === 'TICK') return 'critical';
      if (event.type === 'EXPIRE') return 'expired';
      return state;
    case 'expired':
      if (event.type === 'EXECUTE') return 'executing';
      if (event.type === 'RESET') return 'running';
      return state;
    case 'executing':
      if (event.type === 'EXECUTE_COMPLETE') return 'completed';
      if (event.type === 'EXECUTE_ERROR') return 'expired';
      return state;
    case 'completed':
      return state;
    case 'paused':
      if (event.type === 'RESUME') return 'running';
      return state;
    default:
      return state;
  }
}

/* ---------------------------------------------------------------------------
 * Helpers
 * ------------------------------------------------------------------------- */

interface TimeRemaining { days: number; hours: number; minutes: number; seconds: number; totalMs: number; }

function computeTimeRemaining(deadline: Date): TimeRemaining {
  const totalMs = Math.max(0, deadline.getTime() - Date.now());
  const totalSeconds = Math.floor(totalMs / 1000);
  return {
    days: Math.floor(totalSeconds / 86400),
    hours: Math.floor((totalSeconds % 86400) / 3600),
    minutes: Math.floor((totalSeconds % 3600) / 60),
    seconds: totalSeconds % 60,
    totalMs,
  };
}

function formatRemaining(tr: TimeRemaining): string {
  if (tr.totalMs <= 0) return '0s';
  const parts: string[] = [];
  if (tr.days > 0) parts.push(`${tr.days}d`);
  if (tr.hours > 0) parts.push(`${tr.hours}h`);
  if (tr.minutes > 0) parts.push(`${tr.minutes}m`);
  parts.push(`${tr.seconds}s`);
  return parts.join(' ');
}

/* ---------------------------------------------------------------------------
 * Component
 * ------------------------------------------------------------------------- */

export const TimelockCountdown = defineComponent({
  name: 'TimelockCountdown',
  props: {
    phase: { type: String, required: true },
    deadline: { type: String, required: true },
    elapsed: { type: Number, required: true },
    total: { type: Number, required: true },
    showChallenge: { type: Boolean, default: true },
    warningThreshold: { type: Number, default: 0.8 },
    criticalThreshold: { type: Number, default: 0.95 },
    variant: { type: String, default: 'phase-based' },
  },
  emits: ['execute', 'challenge'],
  setup(props, { slots, emit }) {
    const state = ref<TimelockCountdownState>('running');
    const send = (event: TimelockCountdownEvent) => {
      state.value = timelockCountdownReducer(state.value, event);
    };

    const timeRemaining = ref<TimeRemaining>(computeTimeRemaining(new Date(props.deadline)));
    let intervalId: ReturnType<typeof setInterval> | null = null;

    const deadlineDate = computed(() => new Date(props.deadline));
    const progress = computed(() => props.total <= 0 ? 0 : Math.min(1, Math.max(0, props.elapsed / props.total)));
    const countdownText = computed(() => formatRemaining(timeRemaining.value));
    const formattedDeadline = computed(() => {
      try {
        return deadlineDate.value.toLocaleString(undefined, { dateStyle: 'medium', timeStyle: 'short' });
      } catch { return props.deadline; }
    });
    const displayPhase = computed(() => {
      switch (state.value) {
        case 'expired': return 'Ready to execute';
        case 'executing': return 'Executing...';
        case 'completed': return 'Execution complete';
        case 'paused': return `${props.phase} (paused)`;
        default: return props.phase;
      }
    });

    const stopInterval = () => {
      if (intervalId !== null) { clearInterval(intervalId); intervalId = null; }
    };

    const tick = () => {
      const tr = computeTimeRemaining(deadlineDate.value);
      timeRemaining.value = tr;
      if (tr.totalMs <= 0) { send({ type: 'EXPIRE' }); return; }
      const currentProgress = props.total > 0 ? Math.min(1, props.elapsed / props.total) : 0;
      if (state.value === 'running' && currentProgress >= props.warningThreshold) {
        send({ type: 'WARNING_THRESHOLD' });
      } else if (state.value === 'warning' && currentProgress >= props.criticalThreshold) {
        send({ type: 'CRITICAL_THRESHOLD' });
      } else {
        send({ type: 'TICK' });
      }
    };

    watch(state, (s) => {
      const tickingStates: TimelockCountdownState[] = ['running', 'warning', 'critical'];
      if (!tickingStates.includes(s)) { stopInterval(); return; }
      stopInterval();
      tick();
      intervalId = setInterval(tick, 1000);
    }, { immediate: true });

    // Expire immediately if deadline already past
    onMounted(() => {
      if (deadlineDate.value.getTime() <= Date.now() && state.value === 'running') {
        send({ type: 'EXPIRE' });
      }
    });

    onUnmounted(stopInterval);

    const handleExecute = () => {
      if (state.value === 'expired') { send({ type: 'EXECUTE' }); emit('execute'); }
    };
    const handleChallenge = () => {
      const disabled: TimelockCountdownState[] = ['expired', 'completed', 'executing'];
      if (!disabled.includes(state.value)) emit('challenge');
    };

    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Enter') { e.preventDefault(); handleExecute(); }
      if (e.key === 'c' && !e.ctrlKey && !e.metaKey && !e.altKey) { e.preventDefault(); handleChallenge(); }
    };

    return () => {
      const executeDisabled = state.value !== 'expired';
      const challengeDisabled = state.value === 'expired' || state.value === 'completed' || state.value === 'executing';
      const progressPercent = Math.round(progress.value * 100);

      return h('div', {
        role: 'timer', 'aria-label': `${displayPhase.value}: ${countdownText.value}`,
        'aria-live': 'polite',
        'data-surface-widget': '', 'data-widget-name': 'timelock-countdown',
        'data-part': 'root', 'data-state': state.value, 'data-variant': props.variant,
        onKeydown: handleKeyDown, tabindex: 0,
      }, [
        h('span', { 'data-part': 'phase-label', 'data-state': state.value }, displayPhase.value),
        h('span', {
          'data-part': 'countdown-text', 'data-state': state.value,
          'data-urgency': state.value, 'aria-atomic': 'true',
        }, state.value === 'completed' ? 'Done' : countdownText.value),
        h('span', { 'data-part': 'target-date', 'data-state': state.value }, formattedDeadline.value),
        h('div', {
          'data-part': 'progress-bar', 'data-state': state.value, 'data-progress': progress.value,
          role: 'progressbar', 'aria-valuenow': progressPercent,
          'aria-valuemin': 0, 'aria-valuemax': 100,
          'aria-label': `Timelock progress: ${progressPercent}%`,
        }, [
          h('div', {
            'data-part': 'progress-fill',
            style: { width: `${progressPercent}%`, height: '100%', transition: 'width 0.3s ease' },
          }),
        ]),
        h('button', {
          type: 'button', 'data-part': 'execute-button', 'data-state': state.value,
          'aria-label': 'Execute proposal', 'aria-disabled': executeDisabled,
          disabled: executeDisabled, tabindex: 0, onClick: handleExecute,
        }, state.value === 'executing' ? 'Executing...' : 'Execute'),
        props.showChallenge
          ? h('button', {
              type: 'button', 'data-part': 'challenge-button', 'data-state': state.value,
              'data-visible': 'true', 'aria-label': 'Challenge execution',
              'aria-disabled': challengeDisabled, disabled: challengeDisabled,
              tabindex: 0, onClick: handleChallenge,
            }, 'Challenge')
          : null,
        slots.default ? slots.default() : null,
      ]);
    };
  },
});

export default TimelockCountdown;
